x/dex/types/orders.go

```diff
@@ -3,49 +3,145 @@ package types
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/sei-protocol/sei-chain/utils"
-	"github.com/sei-protocol/sei-chain/utils/datastructures"
 )
 
 type OrderBook struct {
-	Longs  *CachedSortedOrderBookEntries
-	Shorts *CachedSortedOrderBookEntries
-}
-
-func (o *OrderBook) DeepCopy() *OrderBook {
-	return &OrderBook{
-		Longs:  o.Longs.DeepCopy(),
-		Shorts: o.Shorts.DeepCopy(),
-	}
+	Contract ContractAddress
+	Pair     Pair
+	Longs    *CachedSortedOrderBookEntries
+	Shorts   *CachedSortedOrderBookEntries
 }
 
 // entries are always sorted by prices in ascending order, regardless of side
 type CachedSortedOrderBookEntries struct {
-	Entries      []OrderBookEntry
-	DirtyEntries *datastructures.TypedSyncMap[string, OrderBookEntry]
+	CachedEntries []OrderBookEntry
+	currentPtr    int
+
+	loader  func(ctx sdk.Context, startingPriceExclusive sdk.Dec, withLimit bool) []OrderBookEntry
+	setter  func(sdk.Context, OrderBookEntry)
+	deleter func(sdk.Context, OrderBookEntry)
 }
 
-func (c *CachedSortedOrderBookEntries) DeepCopy() *CachedSortedOrderBookEntries {
+func NewCachedSortedOrderBookEntries(
+	loader func(ctx sdk.Context, startingPriceExclusive sdk.Dec, withLimit bool) []OrderBookEntry,
+	setter func(sdk.Context, OrderBookEntry),
+	deleter func(sdk.Context, OrderBookEntry),
+) *CachedSortedOrderBookEntries {
 	return &CachedSortedOrderBookEntries{
-		Entries:      utils.Map(c.Entries, func(e OrderBookEntry) OrderBookEntry { return e.DeepCopy() }),
-		DirtyEntries: c.DirtyEntries.DeepCopy(func(e OrderBookEntry) OrderBookEntry { return e.DeepCopy() }),
+		loader:  loader,
+		setter:  setter,
+		deleter: deleter,
+	}
+}
+
+func (c *CachedSortedOrderBookEntries) load(ctx sdk.Context) {
+	var loaded []OrderBookEntry
+	if len(c.CachedEntries) == 0 {
+		loaded = c.loader(ctx, sdk.ZeroDec(), false)
+	} else {
+		loaded = c.loader(ctx, c.CachedEntries[len(c.CachedEntries)-1].GetOrderEntry().Price, true)
+	}
+	c.CachedEntries = append(c.CachedEntries, loaded...)
+}
+
+func (c *CachedSortedOrderBookEntries) SettleQuantity(ctx sdk.Context, quantity sdk.Dec) (res []ToSettle, settled sdk.Dec) {
+	currentEntry := c.CachedEntries[c.currentPtr].GetOrderEntry()
+
+	if quantity.GTE(currentEntry.Quantity) {
+		res = utils.Map(currentEntry.Allocations, AllocationToSettle)
+		settled = currentEntry.Quantity
+		currentEntry.Quantity = sdk.ZeroDec()
+		currentEntry.Allocations = []*Allocation{}
+		return
+	}
+
+	settled = sdk.ZeroDec()
+	newFirstAllocationIdx := 0
+	for idx, a := range currentEntry.Allocations {
+		postSettle := settled.Add(a.Quantity)
+		if postSettle.LTE(quantity) {
+			settled = postSettle
+			res = append(res, AllocationToSettle(a))
+		} else {
+			newFirstAllocationIdx = idx
+			if settled.Equal(quantity) {
+				break
```

Comment:
we should remove allocations[newFirstAllocationIdx] when all its quantity gets settled here right

Reply:
this is the case where the order is fully settled *before* applying the current allocation. In other words the order book entry isn't exhausted yet - it still at least has this current allocation left